### DISCUSSION TITLE
Help with plugin - why isn't `UIManager:close()` working?

### DISCUSSION THREAD

**Discussion by neilswann80** [koreader/koreader#11591] [Q&A] [answered]
I'm trying to fix a bug in my fork of roygbyte's crossword.koplugin.  I'm trying to close the "Puzzle Library" in library.lua:
https://github.com/neilswann80/crossword.koplugin/blob/main/library.lua#L116

Currently the close command is commented out and in its place a more drastic koreader restart which I'd prefer not to have to use.

Any idea why `UIManager:close(LibraryPage)` doesn't seem to work?  Is the widget name being changed when navigating through puzzles?

I also tried `UIManager:close(self)` which crashes koreader.  

Any suggestions for a solution would be most welcome.

**Comment by Frenzie** [comment]
I can't test it right now, but I think you're basically just overriding and replicating what it does if you leave it alone? Which is to say, what happens when you just get rid of the close_callback and onClose?

**Reply by neilswann80** [reply]
Without the close_callback the close button just moves back one navigation step. So with enough presses (depending on how many times you moved through the puzzle library pages) it will eventually exit. What I'm aiming for is a one-press close button exit ideally.

**Comment by pazos** [comment]
> Any idea why UIManager:close(LibraryPage) doesn't seem to work? Is the widget name being changed when navigating through puzzles?

> I also tried UIManager:close(self) which crashes koreader.

`LibraryPage` is a local variable within method `showDirectoryView` of class `Library`. You're trying to use it on another method `onClose` of the same class.

You can solve it by making `LibraryPage` a field within the `Library` class, and populate, show and close `self.LibraryPage`

**Reply by neilswann80** [reply]
Thanks for explaining that.  It's gone way over my head (I'm only 10 days into to playing with LUA).

I'll do a bit of studying and see if I can workout how to implement your instructions.

**Reply by pazos** [reply]
Welcome to lua, it is fun :)

Anyhow, the solution isn't specific to lua: local variables within functions go out of scope when the function returns. 

Just replace: 

```lua
local LibraryPage
```

with

```lua
self.LibraryPage
```

And use `self.LibraryPage` on UIManager show/close

**Reply by neilswann80** [reply]
You teased me with "just replace"  😄 
I presume I still have to make changes to the library class also?

**Reply by pazos** [reply]
I meant:

```diff
--- a.lua	2024-03-29 16:41:12.126364465 +0100
+++ b.lua	2024-03-29 16:43:21.320689554 +0100
@@ -103,17 +103,16 @@
             end
       })
    end
-   local LibraryPage = KeyValuePage:new{
+   self.LibraryPage = KeyValuePage:new{
       title = "Puzzles",
       kv_pairs = kv_pairs,
       close_callback = function() self:onClose() end,
    }
-   UIManager:show(LibraryPage)
+   UIManager:show(self.LibraryPage)
 end

 function Library:onClose()
-   UIManager:restartKOReader()
-   --UIManager:close(LibraryPage)
+   UIManager:close(self.LibraryPage)
    return true
 end
```

> I presume I still have to make changes to the library class also?

It isn't mandatory but make easier to understand the intention if you define `LibraryPage = nil` on the `Library` table

**Reply by neilswann80** [reply]
I did try this, but it had no effect.  I've tried so many permutations and techniques and nothing seems to work.  

Tried setting a variable in the close_callback area, then test  for the variable's value with an if-else on whether to show or close the LibraryPage... but weirdly the variable value wasn't available outside the function (wasn't a local variable either).

Tried creating a text file and testing for the presence of that also using the same if-else structure, but it won't see the created text file until the next run of the plugin.  That I found super odd.

Must have something to do with the KeyValuePage widget?

**Reply by neilswann80** [reply]
For example why doesn't the below work?  If I uncomment the restart in the close_callback section it restarts on pressing the close button.  As-is there is no restart.  😕 

```
   LibraryPage = KeyValuePage:new{
      title = "Puzzles",
      kv_pairs = kv_pairs,
      close_callback = function()
	  isClose = 1
          --UIManager:restartKOReader()
	  end,
   }
   if isClose == 1 then
     UIManager:restartKOReader()
   else 
     UIManager:show(LibraryPage)
   end
end

return Library
```
Also tried creating a text file in close-callback section, then testing for it in the same if-else section above, but it's not found.  If I restart koreader it's then found.  This I find especially strange as I can see it being created in real time on the device.

EDIT:  once the close-callback is activated by close button press it seems like there's no further interactions from the other functions, output from logger.info commands I place at the start of each function halt as soon as I press close.  Everything that happens from that point on trapped within UIManager and/or KeyValuePage?

**Reply by Frenzie** [reply]
All the rest has already been executed, that's the setup. The callbacks are the only thing that'll be executed at some later point in time.

**Reply by neilswann80** [reply]
Is there a way to use the close_callback option to clear the callbacks so it just exits, rather than cycling back through the callback pages?

**Reply by Frenzie** [reply]
The principle is what @pazos said, but perhaps it's the parent you'd like to close.

**Reply by neilswann80** [reply]
I'd like to close any and all ui instances called by the plugin, dropping back to the main koreader view.

**Comment by neilswann80** [comment, marked as answer]
FYI... this worked... clunky but it did it!

In main.lua:
```diff
function Crossword:getSubMenuItems()
   self:lazyInitialization()
   -- This is the standard menu.
   local sub_menu_items = {
      {
         text = _("Puzzle Library"),
         callback = function()
            self:showLibraryView()
+	    firstView = UIManager:getTopmostVisibleWidget()
         end
      },
```
Then in library.lua:
```diff
+  lastView = UIManager:getTopmostVisibleWidget()
   LibraryPage = KeyValuePage:new{
         title = "Puzzles",
         kv_pairs = kv_pairs,
         close_callback = function()
-          UIManager:close(LibraryPage)
+	   if lastView ~= UIManager:getTopmostVisibleWidget() then
+	      UIManager:close(LibraryPage)
+	   else
+	      while (UIManager:getTopmostVisibleWidget() ~= firstView) do
+		UIManager:close(UIManager:getTopmostVisibleWidget())
+	      end
+	      UIManager:close(UIManager:getTopmostVisibleWidget())
+	      UIManager:setDirty(nil, "full")
+	    end
	 end
   }
   UIManager:show(LibraryPage)
end

return Library
```
